### ISSUE TITLE
Coordinate system for global_orient and body_pose in custom SMPL rotation annotations?

### ISSUE THREAD

**Issue by CamellIyquitous** [IsshikiHugh/HSMR#17]
Hi and thank you for your amazing work!

I'm currently trying to construct my own ground truth training dataset for your model, and I have a question regarding how to correctly prepare the SMPL rotation parameters.

When constructing my own `smpl_params`, especially the joint rotations (`global_orient` and `body_pose` in axis-angle or 6D format), which coordinate frame should they be defined in?

Should the rotations be:
- (A) In the **world coordinate frame** (i.e., whatever coordinate system the SMPL motion was originally in), or
- (B) In the **camera coordinate frame**, corresponding to the camera that captured the input image?

I'm particularly concerned about the `global_orient` of the pelvis/root joint — does it need to be expressed in the camera coordinate system to align with the input image?

Any clarification or best practices on this would be super helpful. Thanks a lot!

**Comment by IsshikiHugh** [comment]
Hi,
thank you for being interested in HSMR!

It should be in the **camera coordinates**.
I highly recommend you visualizing your data in a 3D viewer (like [wis3d](https://github.com/IsshikiHugh/HSMR/blob/6087ce3cd4f546728c3689a2328f8c105c00b7a3/lib/utils/vis/wis3d.py#L75-L118)) as well as our data.
For example, you can manually download one zip file from [HuggingFace](https://huggingface.co/datasets/IsshikiHugh/HSMR-TrainingData), and run the SKEL inference, exporting `skin_verts` and `skin_faces`, visualizing them in the wis3d viewer and checking the forward direction according to the related image.

**Comment by CamellIyquitous** [comment]
Thank you so much for the clarification and the helpful advice! 
I’ll try visualizing with wis3d as suggested. 
Appreciate your quick response!

**Comment by CamellIyquitous** [comment]
Hi, when trying to project 3D joints onto the image for visualization, I ran into a small issue.

Could you please clarify:
- Are the **GT 3D joint coordinates** provided in the GT dataset in the **camera coordinate frame** of the corresponding image?
- If so, what is the **axis convention** for that camera coordinate system?

Just to confirm — is it:
- **Y-up**, **X-right**, **Z-towards the viewer** (i.e., Z points out of the image plane towards the observer's eye)?

Thanks again for your time!

**Comment by IsshikiHugh** [comment]
> Hi, when trying to project 3D joints onto the image for visualization, I ran into a small issue.
> 
> Could you please clarify:
> 
> * Are the **GT 3D joint coordinates** provided in the GT dataset in the **camera coordinate frame** of the corresponding image?
> * If so, what is the **axis convention** for that camera coordinate system?
> 
> Just to confirm — is it:
> 
> * **Y-up**, **X-right**, **Z-towards the viewer** (i.e., Z points out of the image plane towards the observer's eye)?
> 
> Thanks again for your time!

Hi, sorry for confusing you.

It should be:

```
\ z-camera_forward
  \___ x-right
  |
  | y-down
```


https://github.com/user-attachments/assets/75324fee-69b9-439f-979e-f06d060ef1c9

**Comment by CamellIyquitous** [comment]
Thank you for the clarification on the camera coordinate system, that's very helpful.

I want to double-check something important:

I understand that the GT 3D joint coordinates are expressed in the camera coordinate system:
- z-axis points forward into the scene,
- x-axis to the right,
- y-axis downward.

Now, regarding the SMPL rotation parameters:
- Are the `global_orient` (i.e., pelvis orientation) also defined in this **same camera coordinate system** you just mentioned?
- Or are they still in the **SMPL canonical coordinate system**?

As far as I know, in the default SMPL setup:
- The default **`global_orient`** is defined in the **SMPL canonical coordinate system**, where:
  - +Y is up
  - +X is right
  - +Z points **towards the viewer**
- The other body joints in `body_pose` are local relative rotations with respect to their parent joints.
- To align with the camera coordinate system you mentioned earlier (`+z` into the scene, `+x` right, `+y` down), did you **apply a rotation transform to `global_orient`** to convert it from SMPL's canonical coordinate into the mentioned camera coordinate system?

Basically, I want to confirm:  
Are **both the SMPL global orientation and the GT 3D joints** in the camera frame where:  
**+z is camera forward, +x is right, +y is down**?

Thanks a lot for your time and effort!

**Comment by IsshikiHugh** [comment]
> Thank you for the clarification on the camera coordinate system, that's very helpful.
> 
> I want to double-check something important:
> 
> I understand that the GT 3D joint coordinates are expressed in the camera coordinate system:
> 
> * z-axis points forward into the scene,
> * x-axis to the right,
> * y-axis downward.
> 
> Now, regarding the SMPL rotation parameters:
> 
> * Are the `global_orient` (i.e., pelvis orientation) also defined in this **same camera coordinate system** you just mentioned?
> * Or are they still in the **SMPL canonical coordinate system**?
> 
> As far as I know, in the default SMPL setup:
> 
> * The default **`global_orient`** is defined in the **SMPL canonical coordinate system**, where:
>   
>   * +Y is up
>   * +X is right
>   * +Z points **towards the viewer**
> * The other body joints in `body_pose` are local relative rotations with respect to their parent joints.
> * To align with the camera coordinate system you mentioned earlier (`+z` into the scene, `+x` right, `+y` down), did you **apply a rotation transform to `global_orient`** to convert it from SMPL's canonical coordinate into the mentioned camera coordinate system?
> 
> Basically, I want to confirm: Are **both the SMPL global orientation and the GT 3D joints** in the camera frame where: **+z is camera forward, +x is right, +y is down**?
> 
> Thanks a lot for your time and effort!

SMPL parameters should align the GT kp3d. Thus:

> Are both the SMPL global orientation and the GT 3D joints in the camera frame where:
> +z is camera forward, +x is right, +y is down?

This is right. Or, in the 3D viewer, the result SMPL mesh should in the same direction as kp3d.

**Comment by CamellIyquitous** [comment]
It's very useful. Thanks for your time and patience!

**Comment by CamellIyquitous** [comment]
Sorry for reaching out again QAQ...

I still have some confusion regarding how exactly to transform my own GT SMPL parameters (which are currently defined in the canonical SMPL coordinate frame) into the correct OpenCV-style camera coordinate (you mentioned) frame corresponding to the images.

![Image](https://github.com/user-attachments/assets/30b3ef43-f193-40c0-8e01-240e2a9cbe2b)
![Image](https://github.com/user-attachments/assets/31af65dd-a039-4bfb-aa63-b57fdcff1a2e)

Would it be possible for me to reach out to you via a more direct channel (like WeChat? my WeChat: Nacreous_Snowmelt) to ask for some more detailed guidance?  
Of course, only if you're okay with it.

Thank you very much again for your time and help!

**Comment by IsshikiHugh** [comment]
> Would it be possible for me to reach out to you via a more direct channel (like WeChat? my WeChat: Nacreous_Snowmelt) to ask for some more detailed guidance?

Generally I'd suggest you post your questions here so that others who have the same problems can find a solution themselves. But anyway I have sent you the requests.
